x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,684 @@
+
+	for ; iterator.Valid(); iterator.Next() {
+		var lvg types.LocalVirtualGroup
+		k.cdc.MustUnmarshal(iterator.Value(), &lvg)
+		lvgs = append(lvgs, &lvg)
+	}
+
+	return lvgs
+}
+
+func (k Keeper) GetGVGsBindingOnBucket(ctx sdk.Context, bucketID math.Uint) (*types.GlobalVirtualGroupsBindingOnBucket, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGsBindingOnBucketKey(bucketID))
+	if bz == nil {
+		return nil, false
+	}
+
+	var gvgsBindingOnBucket types.GlobalVirtualGroupsBindingOnBucket
+	k.cdc.MustUnmarshal(bz, &gvgsBindingOnBucket)
+	return &gvgsBindingOnBucket, true
+}
+
+func (k Keeper) SetGVGsBindingOnBucket(ctx sdk.Context, gvgsBindingOnBucket *types.GlobalVirtualGroupsBindingOnBucket) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvgsBindingOnBucket)
+	store.Set(types.GetGVGsBindingOnBucketKey(gvgsBindingOnBucket.BucketId), bz)
+}
+
+func (k Keeper) SetGVGFamily(ctx sdk.Context, primarySpID uint32, gvgFamily *types.GlobalVirtualGroupFamily) {
+
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvgFamily)
+	store.Set(types.GetGVGFamilyKey(primarySpID, gvgFamily.Id), bz)
+}
+
+func (k Keeper) GetGVGFamily(ctx sdk.Context, spID, familyID uint32) (*types.GlobalVirtualGroupFamily, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	var gvgFamily types.GlobalVirtualGroupFamily
+	bz := store.Get(types.GetGVGFamilyKey(spID, familyID))
+	if bz == nil {
+		return nil, false
+	}
+	k.cdc.MustUnmarshal(bz, &gvgFamily)
+	return &gvgFamily, true
+}
+
+func (k Keeper) GetOrCreateEmptyGVGFamily(ctx sdk.Context, familyID uint32, spID uint32) (*types.GlobalVirtualGroupFamily, error) {
+	store := ctx.KVStore(k.storeKey)
+	var gvgFamily types.GlobalVirtualGroupFamily
+	// If familyID is not specified, a new family needs to be created
+	if familyID == types.NoSpecifiedFamilyId {
+		id := k.GenNextGVGFamilyID(ctx)
+		gvgFamily = types.GlobalVirtualGroupFamily{
+			Id:                    id,
+			VirtualPaymentAddress: k.DeriveVirtualPaymentAccount(types.GVGFamilyName, id).String(),
+		}
+		return &gvgFamily, nil
+	} else {
+		bz := store.Get(types.GetGVGFamilyKey(spID, familyID))
+		if bz == nil {
+			return nil, types.ErrGVGFamilyNotExist
+		}
+		k.cdc.MustUnmarshal(bz, &gvgFamily)
+
+		return &gvgFamily, nil
+	}
+}
+
+func (k Keeper) DeriveVirtualPaymentAccount(groupType string, id uint32) sdk.AccAddress {
+	b := make([]byte, 4)
+	binary.LittleEndian.PutUint32(b, id)
+
+	return address.Module(types.ModuleName, append([]byte(groupType), b...))
+}
+
+func (k Keeper) GetAvailableStakingTokens(ctx sdk.Context, gvg *types.GlobalVirtualGroup) math.Int {
+	stakingPrice := k.GVGStakingPrice(ctx)
+
+	mustStakingTokens := stakingPrice.MulInt64(int64(gvg.StoredSize))
+
+	return gvg.TotalDeposit.Sub(mustStakingTokens.TruncateInt())
+}
+
+func (k Keeper) BindingObjectToGVG(ctx sdk.Context, bucketID math.Uint, primarySPID, familyID, gvgID uint32, payloadSize uint64) (*types.LocalVirtualGroup, error) {
```

Comment:
It is weird that `BindingObjectToGVG` does not have parameter related with object.